database/migrations/000090_entity_properties.up.sql

```diff
@@ -0,0 +1,51 @@
+-- Copyright 2024 Stacklok, Inc
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--      http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS entities (
+	id          UUID NOT NULL DEFAULT gen_random_uuid() PRIMARY KEY,
+	entity_type entities NOT NULL, 
+	name        TEXT NOT NULL,
+	project_id  UUID NOT NULL REFERENCES project(id),
+	provider_id UUID NOT NULL REFERENCES provider(id),
+	created_at  TIMESTAMP NOT NULL DEFAULT NOW(),
+	most_recent_evaluation TIMESTAMP, -- last time a profile/rule was run against this entity
+	UNIQUE(project_id, provider_id, entity_type, name)
+);
+
+CREATE TABLE IF NOT EXISTS properties(
+	id          UUID PRIMARY KEY, -- surrogate ID
+	entity_id   UUID NOT NULL REFERENCES entities(id) ON DELETE CASCADE,
+	key         TEXT NOT NULL, -- we need to validate and ensure there are no dots
+	value       JSONB NOT NULL,
+	updated_at  TIMESTAMP NOT NULL DEFAULT NOW(),
+	UNIQUE (entity_id, key)
+);
+
+COMMIT;
+
+-- Populate the properties table with the existing data. Note we don't
+-- have to worry about the most_recent_evaluation column because unused still.
+-- Also note that we're not populating properties just yet. We'll do that
+-- in code as we need to.
+
+INSERT INTO entities (id, entity_type, name, project_id, provider_id, created_at)
+SELECT id, 'repository', name, project_id, provider_id, created_at FROM repositories;
+
+INSERT INTO entities (id, entity_type, name, project_id, provider_id, created_at)
+SELECT id, 'artifact', name, project_id, provider_id, created_at FROM artifacts;
+
+INSERT INTO entities (id, entity_type, name, project_id, provider_id, created_at)
+SELECT id, 'pull_request', name, project_id, provider_id, created_at FROM pull_requests;
```

Comment:
I'm not sure this is enough. Our infra first migrates the DB while the old code is running and then upgrades the minder code, which creates a race where an entry might be added to the old tables but never to the new ones.

What if we create a SQL trigger that starts writing to the new table on create and update instead, commit that first so that the existing code writes to the new tables and then copy the existing data? (I also wonder about the size of the data and if we'd create a huge transaction, but that's a matter of checking our SaaS data size)

Reply:
I think we should just choose a migration on the golang migration wrapper.

Reply:
+1 to doing the migration in the golang wrapper in a later PR after the tables are created and we start writing to them from the application code (but before we assume they are there on the read side).

I worry about doing SQL bulk inserts / updates leading to database locks, though I acknowledge that we haven't had those problems yet.